3.0/m_roleplay.cpp

```diff
@@ -717,6 +803,40 @@ class ModuleRoleplay : public Module
 			ServerInstance->SNO->WriteToSnoMask('a', "WARNING: Roleplay configuration has needchanmode and needop both disabled, this could allow for apparent spoofing!");
 	}
 
+	ModResult OnUserPreMessage(User* user, const MessageTarget& target, MessageDetails& details) CXX11_OVERRIDE
+	{
+		return CopyRoleplayTags(details.tags_in, details.tags_out);
+	}
+
+	ModResult OnUserPreTagMessage(User* user, const MessageTarget& target, CTCTags::TagMessageDetails& details) CXX11_OVERRIDE
+	{
+		return CopyRoleplayTags(details.tags_in, details.tags_out);
+	}
+
+	// This is where the magic of rewriting the user happens.
+	ModResult OnUserWrite(LocalUser* user, ClientProtocol::Message& msg) CXX11_OVERRIDE
+	{
+		const ClientProtocol::TagMap& tags = msg.GetTags();
+		ClientProtocol::TagMap::const_iterator tag;
+
+		tag = tags.find("inspircd.org/roleplay-src");
+		if(tag == tags.end())
+			return MOD_RES_PASSTHRU;
+
+		const std::string& src = tag->second.value;
+		if(src.empty())
+			// wtf?
+			throw ModuleException("Empty roleplay source tag received");
```

Comment:
Throwing here is potentially usafe, return MOD_RES_DENY if a message is invalid.

Reply:
Done, and added logging.